services/token_airdrop.proto

```diff
@@ -0,0 +1,92 @@
+ *    and must claim that airdrop with a `claimAirdrop` transaction.<br/>
+ *    The fee for this transfer SHALL include the transfer, the association, the cost to renew
+ *    that association once, the airdrop fee, and any custom fees. If the pending airdrop is not
+ *    claimed immediately, the `sender` SHALL pay the cost to renew the token association, and
+ *    the cost to maintain the pending airdrop, until the pending airdrop is claimed or cancelled.
+ *  - An account with no available automatic association slots SHALL have a "Pending Airdrop"
+ *    created and must claim that airdrop with a `claimAirdrop` transaction.<br/>
+ *    The fee for this transfer SHALL include the transfer, the association, the cost to renew
+ *    that association once, the airdrop fee, and any custom fees. If the pending airdrop is not
+ *    claimed immediately, the `sender` SHALL pay the cost to renew the token association, and
+ *    the cost to maintain the pending airdrop, until the pending airdrop is claimed or cancelled.
+ *
+ * If an airdrop would create a pending airdrop for a fungible/common token, and a pending airdrop
+ * for the same sender, receiver, and token already exists, the existing pending airdrop
+ * SHALL be updated to add the new amount to the existing airdrop, rather than creating a new
+ * pending airdrop.
+ *
+ * Any airdrop that completes immediately SHALL be irreversible. Any airdrop that results in a
+ * "Pending Airdrop" MAY be canceled via a `cancelAirdrop` transaction.
+ *
+ * All transfer fees (including custom fees and royalties), as well as the rent cost for the
+ * first auto-renewal period for any automatic-association slot occupied by the airdropped
+ * tokens, SHALL be charged to the account paying for this transaction.
+ *
+ * ### Record Stream Effects
+ * - Each successful transfer SHALL populate that transfer in `token_transfer_list` for the record.
+ * - Each successful transfer that consumes an automatic association slot SHALL populate the
+ *   `automatic_association` field for the record.
+ * - Each pending transfer _created_ SHALL be added to the `pending_airdrops` field for the record.
+ * - Each pending transfer _updated_ SHALL be added to the `pending_airdrops` field for the record.
+ */
+message TokenAirdropTransactionBody {
+  /**
+   * A list of token transfers representing one or more airdrops.<br/>
+   * The sender for each transfer MUST have sufficient balance to complete the transfers.
+   * <p>
+   * All token transfers MUST successfully transfer tokens or create a pending airdrop
+   * for this transaction to succeed.<br/>
+   * This list MUST contain between 1 and 10 transfers, inclusive.
+   * <p>
+   * <blockquote>Note that each transfer of fungible/common tokens requires both a debit and
+   * a credit, so each _fungible_ token transfer MUST have _balanced_ entries in the
+   * TokenTransferList for that transfer.</blockquote>
+   */
```

Comment:
So we are re-using the same message and enforcing this restriction in code ?

Reply:
I hope we already are enforcing correct double-entry records for token transfers.

Reply:
But the problem is this should be mapped to a single sender in `PendingAirdropId`. If there are multiple debits and multiple credits, it is not clear to map to `PendingAirdropId`

Reply:
Each Pending Airdrop is created individually from a single transfer in the list, and only if necessary.  The majority of transfers are expected to complete without creating a pending airdrop.
The entire airdrop is never "pending", only individual transfers.
There is no requirement to map this transfer list directly to each `PendingAirdropId`, as far as I know.